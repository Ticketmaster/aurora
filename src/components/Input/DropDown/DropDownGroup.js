import React, { Fragment } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { Transition } from "react-transition-group";

import { TWO_SIZE_VARIANT } from "../../../utils/sizes";
import { Provider as DropDownProvider } from "./Context";
import SelectionProvider from "../../SelectionProvider/Provider";
import { Consumer as SelectionConsumer } from "../../SelectionProvider/Context";
import {
  ARROWUP,
  ARROWDOWN,
  SPACEBAR,
  ESCAPE,
  TAB
} from "../../../utils/keyCharCodes";
import {
  StyledGroup,
  StyledChildWrapper,
  StyledGroupWrapper,
  StyledChevron,
  StyledSelectedText,
  StyledKeyboardProvider,
  HiddenLabel
} from "./DropDownGroup.styles";

class DropDownGroup extends React.Component {
  componentDidMount() {
    document.addEventListener("click", this.handleOutsideClick);
  }

  static getDerivedStateFromProps(props, state) {
    if (props.isOpen !== state.isOpenPrevProp) {
      return {
        isOpen: props.isOpen,
        isOpenPrevProp: props.isOpen
      };
    }
    return null;
  }

  componentDidUpdate(prevProps, prevState) {
    const { isOpen } = this.state;
    // scroll dropdown to top after opening
    if (!this.props.isOpen && isOpen && isOpen !== prevState.isOpen) {
      this.styledChildWrapper.current.scrollTop = 0;
    }
  }

  componentWillUnmount() {
    document.removeEventListener("click", this.handleOutsideClick);
  }

  onClick = () => {
    this.toggleDropdown();
  };

  onKeyDown = e => {
    const { keyCode } = e;
    const { isOpen } = this.state;

    switch (keyCode) {
      case ESCAPE:
        this.closeDropdown();
        break;
      case TAB:
        if (isOpen) {
          e.preventDefault();
        }
        break;
      case ARROWUP:
      case ARROWDOWN:
        e.preventDefault();
        this.openDropdown();
        break;
      case SPACEBAR:
        e.preventDefault();
        this.toggleDropdown();
        break;
      default:
        break;
    }
  };

  getCurrentSelection = value => {
    const selectedItem = React.Children.toArray(this.props.children).find(
      c => c.props.value === value
    );

    return selectedItem && selectedItem.props.children;
  };

  closeDropdown = () => this.setState({ isOpen: false });

  openDropdown = () => this.setState({ isOpen: true });

  toggleDropdown = () => {
    if (this.state.isOpen) {
      this.closeDropdown();
    } else {
      this.openDropdown();
    }
  };

  stopInteraction = e => {
    e.preventDefault();
    e.stopPropagation();
  };

  handleOutsideClick = e => {
    if (this.state.isOpen && !this.groupWrapper.current.contains(e.target)) {
      this.closeDropdown();
    }
  };

  displayLabel = selected => {
    const { placeholder, label } = this.props;

    if (placeholder.length > 0 && selected.length === 0) {
      return placeholder;
    }

    if (selected.length > 0 && label.length > 0) {
      return (
        <Fragment>
          {label} {this.getCurrentSelection(selected[0])}
        </Fragment>
      );
    }

    return this.getCurrentSelection(selected[0]);
  };

  groupWrapper = React.createRef();
  styledChildWrapper = React.createRef();
  ANIMATION_TIMEOUT = 300;

  /* eslint-disable */
  state = {
    isOpen: false,
    isOpenPrevProp: false,
    onClose: this.onClick
  };
  /* eslint-enable */

  render() {
    const {
      children,
      value,
      valueOverride,
      onChange,
      variant,
      isOpen: isOpenProp,
      keywordSearch,
      withKeyboardProvider,
      placeholder,
      label,
      disabled,
      size,
      shouldOpenDownward,
      ...props
    } = this.props;
    const { isOpen: isOpenState } = this.state;
    const isOpen = isOpenProp || isOpenState;
    const hiddenLabelId = `hidden-label__${(placeholder || label).replace(
      / /g,
      "_"
    )}`;
    const onClickListener = disabled
      ? { onMouseDown: this.stopInteraction }
      : { onClick: this.onClick };

    return (
      <SelectionProvider
        onChange={onChange}
        isMultiSelect={false}
        value={value}
        valueOverride={valueOverride}
      >
        <SelectionConsumer>
          {({ selected }) => {
            const keyboardProviderProps = withKeyboardProvider
              ? { keywordSearch, selected }
              : { as: "div" };

            return (
              <Transition in={isOpen} timeout={this.ANIMATION_TIMEOUT}>
                {state => {
                  // keep dropdown--open-upward className until collapse animation is finished (.3s)
                  const hasOpenUpwardClass =
                    state !== "exited" && !shouldOpenDownward;

                  return (
                    <StyledGroupWrapper
                      {...props}
                      className={classNames(props.className, {
                        "dropdown--open-upward": hasOpenUpwardClass,
                        "dropdown--disabled": disabled
                      })}
                      tabIndex={disabled ? -1 : 0}
                      aria-haspopup="listbox"
                      aria-labelledby={hiddenLabelId}
                      onKeyDown={this.onKeyDown}
                      ref={this.groupWrapper}
                    >
                      <StyledGroup
                        {...onClickListener}
                        className={classNames(`dropdown--${size}`, {
                          "dropdown--active": isOpen,
                          "dropdown--border": variant === 0,
                          "dropdown--no-border": variant === 1,
                          "dropdown__label--disabled": disabled
                        })}
                      >
                        {/* HiddenLabel is required for correct screen readers 
                          readings when an option is selected */}
                        <HiddenLabel id={hiddenLabelId}>
                          {placeholder || label}
                        </HiddenLabel>
                        <StyledSelectedText
                          className={classNames({
                            "dropdown__text--disabled": disabled
                          })}
                        >
                          {this.displayLabel(selected)}
                        </StyledSelectedText>

                        <StyledChevron
                          className={classNames({
                            "dropdown__icon--hide": isOpen,
                            "dropdown--no-border": variant === 1,
                            "dropdown__chevron--disabled": disabled
                          })}
                        />
                      </StyledGroup>
                      <Transition in={isOpen} timeout={this.ANIMATION_TIMEOUT}>
                        {wrapperState => (
                          <StyledChildWrapper
                            className={classNames(
                              "dropdown__items",
                              `dropdown__items--${size}`,
                              {
                                "dropdown--clicked": isOpen,
                                "dropdown--overflow": wrapperState === "entered"
                              }
                            )}
                            ref={this.styledChildWrapper}
                          >
                            <DropDownProvider value={{ ...this.state, isOpen }}>
                              <StyledKeyboardProvider
                                role="listbox"
                                aria-labelledby={hiddenLabelId}
                                {...keyboardProviderProps}
                              >
                                {children}
                              </StyledKeyboardProvider>
                            </DropDownProvider>
                          </StyledChildWrapper>
                        )}
                      </Transition>
                    </StyledGroupWrapper>
                  );
                }}
              </Transition>
            );
          }}
        </SelectionConsumer>
      </SelectionProvider>
    );
  }
}

DropDownGroup.propTypes = {
  value: PropTypes.arrayOf(PropTypes.string),
  valueOverride: PropTypes.arrayOf(PropTypes.string),
  children: PropTypes.node.isRequired,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  variant: PropTypes.number,
  label: PropTypes.string,
  isOpen: PropTypes.bool,
  keywordSearch: PropTypes.bool,
  withKeyboardProvider: PropTypes.bool,
  disabled: PropTypes.bool,
  size: PropTypes.oneOf(TWO_SIZE_VARIANT),
  shouldOpenDownward: PropTypes.bool
};

DropDownGroup.defaultProps = {
  value: [],
  valueOverride: null,
  onChange: null,
  placeholder: "",
  variant: 0,
  isOpen: false,
  keywordSearch: true,
  withKeyboardProvider: true,
  label: "",
  disabled: false,
  size: TWO_SIZE_VARIANT[1],
  shouldOpenDownward: true
};

export default DropDownGroup;
