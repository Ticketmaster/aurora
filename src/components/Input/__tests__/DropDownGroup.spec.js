import React from "react";
import renderer from "react-test-renderer";
import {
  renderIntoDocument,
  cleanup,
  fireEvent,
  Simulate
} from "react-testing-library";

import DropDownGroup from "../DropDown/DropDownGroup";
import DropDownOption from "../DropDown/DropDownOption";

describe("DropDownGroup", () => {
  function renderTestComponentOne(props = {}) {
    return renderIntoDocument(
      <div data-testid="test-outSideComponent">
        <div data-testid="something">something</div>
        <DropDownGroup {...props} data-testid="test-dropContainer">
          <DropDownOption
            data-testid="test-dropDownOptionOne"
            value="ValueOne"
            index={0}
          >
            Option One
          </DropDownOption>
          <DropDownOption
            data-testid="test-dropDownOptionTwo"
            value="ValueTwo"
            index={1}
          >
            Second Option
          </DropDownOption>
          <DropDownOption
            data-testid="test-dropDownOptionThree"
            value="ValueThree"
            index={2}
          >
            Option Three
          </DropDownOption>
        </DropDownGroup>
      </div>
    );
  }

  function renderTestComponentTwo(props = {}) {
    return renderIntoDocument(
      <DropDownGroup {...props} variant={1} label="Sort By:">
        <DropDownOption value="date" index={0}>
          {"Date"}
          {null}
        </DropDownOption>
        <DropDownOption value="distance" index={1}>
          {"Distance - "}
          {"Closest"}
        </DropDownOption>
        <DropDownOption value="price" index={2}>
          {"Price "}
          {<span>On sale</span>}
        </DropDownOption>
      </DropDownGroup>
    );
  }

  afterEach(cleanup);

  it("renders default dropdown", () => {
    expect(renderTestComponentOne().container.firstChild).toMatchSnapshot();
  });

  it("renders default dropdown that should always open downwards", () => {
    expect(
      renderTestComponentOne({ shouldOpenDownward: true }).container.firstChild
    ).toMatchSnapshot();
  });

  it("renders small dropdown", () => {
    expect(
      renderTestComponentOne({ size: "small" }).container.firstChild
    ).toMatchSnapshot();
  });

  it("renders variant 0 with a placeholder prop", () => {
    expect(
      renderTestComponentOne({ placeholder: "Select An Option", variant: 1 })
        .container.firstChild
    ).toMatchSnapshot();
  });

  it("renders variant 1 with a label prop", () => {
    expect(
      renderTestComponentOne({ label: "Selected Option:", variant: 1 })
        .container.firstChild
    ).toMatchSnapshot();
  });

  it("renders input correctly when the isOpen prop with a value of true is passed", () => {
    expect(
      renderTestComponentOne({ isOpen: true }).container.firstChild
    ).toMatchSnapshot();
  });

  it("renders input correctly when the keywordSearch prop with a value of false is passed", () => {
    expect(
      renderTestComponentOne({ keywordSearch: false }).container.firstChild
    ).toMatchSnapshot();
  });

  it("renders input correctly when the withKeyboardProvider prop with a value of false is passed", () => {
    expect(
      renderTestComponentOne({ withKeyboardProvider: false }).container
        .firstChild
    ).toMatchSnapshot();
  });

  it("Should open the drop down onClick", () => {
    const { container, getByTestId } = renderTestComponentOne();

    fireEvent.click(getByTestId("test-dropContainer"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Should close the drop down on click outside", () => {
    const { container, getByTestId } = renderTestComponentOne();

    fireEvent.click(getByTestId("test-dropContainer"));
    fireEvent.click(getByTestId("something"));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Should not focus onClick", () => {
    const { container } = renderTestComponentOne({ disabled: true });
    fireEvent.mouseDown(container.querySelector("label"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Escape key should close the drop down", () => {
    const { container, getByTestId } = renderTestComponentOne();

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "Escape",
      keyCode: 27,
      which: 27
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Arrow key up should open the drop down", () => {
    const { container, getByTestId } = renderTestComponentOne();

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "ArrowUp",
      keyCode: 32,
      which: 32
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Arrow down arrow should open the drop down", () => {
    const { container, getByTestId } = renderTestComponentOne();

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "ArrowDown",
      keyCode: 40,
      which: 40
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Click outside should close the dropdown", () => {
    const { container, getByTestId } = renderTestComponentOne();

    fireEvent.click(getByTestId("test-dropContainer"));
    fireEvent.click(getByTestId("test-outSideComponent"));
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Click outside should not close the dropdown when the isOpen prop equals true", () => {
    const { container, getByTestId } = renderTestComponentOne({ isOpen: true });

    fireEvent.click(getByTestId("test-dropDownOptionOne"));
    fireEvent.click(getByTestId("test-outSideComponent"));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("Space bar should select and close dropdown", () => {
    const onChange = jest.fn();
    const { container, getByTestId } = renderTestComponentOne({ onChange });

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "ArrowDown",
      keyCode: 40,
      which: 40,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "",
      keyCode: 32,
      which: 32,
      bubbles: true
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Space bar should select and not close dropdown when the isOpen prop equals true", () => {
    const onChange = jest.fn();
    const { container, getByTestId } = renderTestComponentOne({
      onChange,
      isOpen: true
    });

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "ArrowDown",
      keyCode: 40,
      which: 40,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "",
      keyCode: 32,
      which: 32,
      bubbles: true
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Should remove event listener when unmount", () => {
    const { container, unmount } = renderTestComponentOne();

    unmount();
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("Should invoke an onChange handler option selected", () => {
    const onChange = jest.fn();
    const { getByTestId } = renderTestComponentOne({ onChange });

    Simulate.click(getByTestId("test-dropContainer"));
    Simulate.click(getByTestId("test-dropDownOptionOne"));

    expect(onChange).toHaveBeenCalledWith(["ValueOne"]);
  });

  it("Should invoke an onChange handler when passed value to group", () => {
    const { queryAllByText } = renderTestComponentOne({
      value: ["ValueOne"]
    });

    const optionChoice = queryAllByText("Option One");
    expect(optionChoice).toHaveLength(2);
  });

  it("Selecting options after passing in values", () => {
    const { queryAllByText, getByTestId } = renderTestComponentOne({
      value: ["ValueOne"]
    });

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "ArrowDown",
      keyCode: 40,
      which: 40,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "ArrowDown",
      keyCode: 40,
      which: 40,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "",
      keyCode: 32,
      which: 32,
      bubbles: true
    });

    const optionChoice = queryAllByText("Second Option");
    expect(optionChoice).toHaveLength(1);
  });

  it("Pressing tab when the dropdown is open", () => {
    const { getByTestId } = renderTestComponentOne();
    const preventDefault = jest.fn();

    fireEvent.click(getByTestId("test-dropDownOptionOne"));

    Simulate.keyDown(getByTestId("test-dropContainer"), {
      key: "Tab",
      keyCode: 9,
      which: 9,
      bubbles: true,
      preventDefault
    });

    expect(preventDefault).toHaveBeenCalled();
  });

  it("Should select options based on typed input", () => {
    const { queryAllByText, getByTestId } = renderTestComponentOne();

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "ArrowDown",
      keyCode: 40,
      which: 40,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "s",
      keyCode: 83,
      which: 83,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "",
      keyCode: 32,
      which: 32,
      bubbles: true
    });

    const optionChoice = queryAllByText("Second Option");
    expect(optionChoice).toHaveLength(1);
  });

  it("if entered texted doesn't not match then select the option that matched previously", () => {
    const onChange = jest.fn();
    const { queryAllByText, getByTestId } = renderTestComponentOne({
      onChange
    });

    fireEvent.keyDown(getByTestId("test-dropContainer"), {
      key: "ArrowDown",
      keyCode: 40,
      which: 40,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "s",
      keyCode: 83,
      which: 83,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "z",
      keyCode: 90,
      which: 90,
      bubbles: true
    });

    fireEvent.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "",
      keyCode: 32,
      which: 32,
      bubbles: true
    });

    const optionChoice = queryAllByText("Option One");
    expect(optionChoice).toHaveLength(2);
  });

  it("should NOT disable TAB key when the dropdown is NOT open", () => {
    const preventDefault = jest.fn();
    const { getByTestId } = renderTestComponentOne();

    Simulate.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "",
      keyCode: 9,
      which: 9,
      bubbles: true,
      preventDefault
    });
    expect(preventDefault).not.toBeCalled();
  });

  it("should disable TAB key when the dropdown is open", () => {
    const preventDefault = jest.fn();
    const { getByTestId } = renderTestComponentOne();
    fireEvent.click(getByTestId("test-dropDownOptionOne"));

    Simulate.keyDown(getByTestId("test-dropDownOptionOne"), {
      key: "",
      keyCode: 9,
      which: 9,
      bubbles: true,
      preventDefault
    });
    expect(preventDefault).toHaveBeenCalled();
  });

  it("should disable scroll when dropdown is open", () => {
    const { container, getByTestId } = renderTestComponentOne();

    fireEvent.click(getByTestId("test-dropDownOptionOne"));
    fireEvent.wheel(getByTestId("test-outSideComponent"));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should allow scroll inside the dropdown when it is open", () => {
    const { container, getByTestId } = renderTestComponentOne();

    fireEvent.click(getByTestId("test-dropDownOptionOne"));
    fireEvent.wheel(getByTestId("test-dropDownOptionOne"));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render the correct label when variant equals 1, a label prop is passed, and when an array of children containing falsy values is passed to each DropDownOption", () => {
    const { container } = renderTestComponentTwo({ value: ["date"] });
    expect(container).toMatchSnapshot();
  });

  it("should render the correct label when variant equals 1, a label prop is passed, and when an array of children containing string values is passed to each DropDownOption", () => {
    const { container } = renderTestComponentTwo({ value: ["distance"] });
    expect(container).toMatchSnapshot();
  });

  it("should render the correct label when variant equals 1, a label prop is passed, and when an array of children containing string and node values is passed to each DropDownOption", () => {
    const { container } = renderTestComponentTwo({ value: ["price"] });
    expect(container).toMatchSnapshot();
  });
});

describe("DropDownOption", () => {
  it("should render the correct markup when a className prop is passed", () => {
    const component = renderer.create(
      <DropDownOption
        className="dropdown__item--expandable"
        value="ValueOne"
        index={0}
      >
        Option One
      </DropDownOption>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
