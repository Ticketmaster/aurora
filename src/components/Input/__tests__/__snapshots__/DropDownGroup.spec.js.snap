// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownGroup Arrow down arrow should open the drop down 1`] = `
<div
  data-testid="test-outSideComponent"
>
  <div>
    something
  </div>
  <div
    aria-haspopup="listbox"
    class="sc-htpNat bEXorH"
    data-testid="test-dropContainer"
    label="Sort By:"
    placeholder=""
    tabindex="0"
  >
    <label
      class="dropdown--active dropdown--border sc-bdVaJa jOVSaQ"
    >
      <div
        class="dropdown--selected dropdown--border sc-ifAKCX dIQjbB"
      />
      <svg
        class="dropdown__icon--hide sc-bxivhb bfTkLM"
        fill="currentColor"
        height="12"
        viewBox="0 -1 15 7"
        width="12"
      >
        <path
          d="M13.974.132a.614.614 0 0 0-.762 0L7.17 5.341 1.12.132C.916-.044.563-.037.376.125a.398.398 0 0 0-.167.332c0 .058.012.116.036.17a.385.385 0 0 0 .12.155l6.427 5.54a.568.568 0 0 0 .378.135.568.568 0 0 0 .377-.135l6.427-5.54a.422.422 0 0 0 .156-.325.42.42 0 0 0-.156-.325z"
        />
      </svg>
    </label>
    <div
      class="dropdown__items dropdown--clicked sc-bwzfXH ewSIUN"
    >
      <div
        class="sc-EHOje hRvppO"
        role="listbox"
      >
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionOne"
          tabindex="-1"
          value="ValueOne"
        >
          Option One
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionTwo"
          tabindex="-1"
          value="ValueTwo"
        >
          Second Option
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionThree"
          tabindex="-1"
          value="ValueThree"
        >
          Option Three
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropDownGroup Arrow key up should open the drop down 1`] = `
<div
  data-testid="test-outSideComponent"
>
  <div>
    something
  </div>
  <div
    aria-haspopup="listbox"
    class="sc-htpNat bEXorH"
    data-testid="test-dropContainer"
    label="Sort By:"
    placeholder=""
    tabindex="0"
  >
    <label
      class="dropdown--active dropdown--border sc-bdVaJa jOVSaQ"
    >
      <div
        class="dropdown--selected dropdown--border sc-ifAKCX dIQjbB"
      />
      <svg
        class="dropdown__icon--hide sc-bxivhb bfTkLM"
        fill="currentColor"
        height="12"
        viewBox="0 -1 15 7"
        width="12"
      >
        <path
          d="M13.974.132a.614.614 0 0 0-.762 0L7.17 5.341 1.12.132C.916-.044.563-.037.376.125a.398.398 0 0 0-.167.332c0 .058.012.116.036.17a.385.385 0 0 0 .12.155l6.427 5.54a.568.568 0 0 0 .378.135.568.568 0 0 0 .377-.135l6.427-5.54a.422.422 0 0 0 .156-.325.42.42 0 0 0-.156-.325z"
        />
      </svg>
    </label>
    <div
      class="dropdown__items dropdown--clicked sc-bwzfXH ewSIUN"
    >
      <div
        class="sc-EHOje hRvppO"
        role="listbox"
      >
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionOne"
          tabindex="-1"
          value="ValueOne"
        >
          Option One
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionTwo"
          tabindex="-1"
          value="ValueTwo"
        >
          Second Option
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionThree"
          tabindex="-1"
          value="ValueThree"
        >
          Option Three
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropDownGroup Click outside should close the dropdown 1`] = `
<div
  data-testid="test-outSideComponent"
>
  <div>
    something
  </div>
  <div
    aria-haspopup="listbox"
    class="sc-htpNat bEXorH"
    data-testid="test-dropContainer"
    label="Sort By:"
    placeholder=""
    tabindex="0"
  >
    <label
      class="dropdown--border sc-bdVaJa jOVSaQ"
    >
      <div
        class="dropdown--selected dropdown--border sc-ifAKCX dIQjbB"
      />
      <svg
        class="sc-bxivhb bfTkLM"
        fill="currentColor"
        height="12"
        viewBox="0 -1 15 7"
        width="12"
      >
        <path
          d="M13.974.132a.614.614 0 0 0-.762 0L7.17 5.341 1.12.132C.916-.044.563-.037.376.125a.398.398 0 0 0-.167.332c0 .058.012.116.036.17a.385.385 0 0 0 .12.155l6.427 5.54a.568.568 0 0 0 .378.135.568.568 0 0 0 .377-.135l6.427-5.54a.422.422 0 0 0 .156-.325.42.42 0 0 0-.156-.325z"
        />
      </svg>
    </label>
    <div
      class="dropdown__items sc-bwzfXH ewSIUN"
    >
      <div
        class="sc-EHOje hRvppO"
        role="listbox"
      >
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionOne"
          tabindex="-1"
          value="ValueOne"
        >
          Option One
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionTwo"
          tabindex="-1"
          value="ValueTwo"
        >
          Second Option
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionThree"
          tabindex="-1"
          value="ValueThree"
        >
          Option Three
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropDownGroup Should open the drop down onClick 1`] = `
<div
  data-testid="test-outSideComponent"
>
  <div>
    something
  </div>
  <div
    aria-haspopup="listbox"
    class="sc-htpNat bEXorH"
    data-testid="test-dropContainer"
    label="Sort By:"
    placeholder=""
    tabindex="0"
  >
    <label
      class="dropdown--border sc-bdVaJa jOVSaQ"
    >
      <div
        class="dropdown--selected dropdown--border sc-ifAKCX dIQjbB"
      />
      <svg
        class="sc-bxivhb bfTkLM"
        fill="currentColor"
        height="12"
        viewBox="0 -1 15 7"
        width="12"
      >
        <path
          d="M13.974.132a.614.614 0 0 0-.762 0L7.17 5.341 1.12.132C.916-.044.563-.037.376.125a.398.398 0 0 0-.167.332c0 .058.012.116.036.17a.385.385 0 0 0 .12.155l6.427 5.54a.568.568 0 0 0 .378.135.568.568 0 0 0 .377-.135l6.427-5.54a.422.422 0 0 0 .156-.325.42.42 0 0 0-.156-.325z"
        />
      </svg>
    </label>
    <div
      class="dropdown__items sc-bwzfXH ewSIUN"
    >
      <div
        class="sc-EHOje hRvppO"
        role="listbox"
      >
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionOne"
          tabindex="-1"
          value="ValueOne"
        >
          Option One
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionTwo"
          tabindex="-1"
          value="ValueTwo"
        >
          Second Option
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionThree"
          tabindex="-1"
          value="ValueThree"
        >
          Option Three
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropDownGroup Should remove event listener when unmount 1`] = `""`;

exports[`DropDownGroup Space bar should select and close dropdown 1`] = `
<div
  data-testid="test-outSideComponent"
>
  <div>
    something
  </div>
  <div
    aria-haspopup="listbox"
    class="sc-htpNat bEXorH"
    data-testid="test-dropContainer"
    label="Sort By:"
    placeholder=""
    tabindex="0"
  >
    <label
      class="dropdown--border sc-bdVaJa jOVSaQ"
    >
      <div
        class="dropdown--selected dropdown--border sc-ifAKCX dIQjbB"
      >
        Option One
      </div>
      <svg
        class="sc-bxivhb bfTkLM"
        fill="currentColor"
        height="12"
        viewBox="0 -1 15 7"
        width="12"
      >
        <path
          d="M13.974.132a.614.614 0 0 0-.762 0L7.17 5.341 1.12.132C.916-.044.563-.037.376.125a.398.398 0 0 0-.167.332c0 .058.012.116.036.17a.385.385 0 0 0 .12.155l6.427 5.54a.568.568 0 0 0 .378.135.568.568 0 0 0 .377-.135l6.427-5.54a.422.422 0 0 0 .156-.325.42.42 0 0 0-.156-.325z"
        />
      </svg>
    </label>
    <div
      class="dropdown__items sc-bwzfXH ewSIUN"
    >
      <div
        class="sc-EHOje hRvppO"
        role="listbox"
      >
        <span
          class="dropdown__selected sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionOne"
          tabindex="-1"
          value="ValueOne"
        >
          Option One
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionTwo"
          tabindex="-1"
          value="ValueTwo"
        >
          Second Option
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionThree"
          tabindex="-1"
          value="ValueThree"
        >
          Option Three
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropDownGroup renders default input 1`] = `
<div
  data-testid="test-outSideComponent"
>
  <div>
    something
  </div>
  <div
    aria-haspopup="listbox"
    class="sc-htpNat bEXorH"
    data-testid="test-dropContainer"
    label="Sort By:"
    placeholder=""
    tabindex="0"
  >
    <label
      class="dropdown--border sc-bdVaJa jOVSaQ"
    >
      <div
        class="dropdown--selected dropdown--border sc-ifAKCX dIQjbB"
      />
      <svg
        class="sc-bxivhb bfTkLM"
        fill="currentColor"
        height="12"
        viewBox="0 -1 15 7"
        width="12"
      >
        <path
          d="M13.974.132a.614.614 0 0 0-.762 0L7.17 5.341 1.12.132C.916-.044.563-.037.376.125a.398.398 0 0 0-.167.332c0 .058.012.116.036.17a.385.385 0 0 0 .12.155l6.427 5.54a.568.568 0 0 0 .378.135.568.568 0 0 0 .377-.135l6.427-5.54a.422.422 0 0 0 .156-.325.42.42 0 0 0-.156-.325z"
        />
      </svg>
    </label>
    <div
      class="dropdown__items sc-bwzfXH ewSIUN"
    >
      <div
        class="sc-EHOje hRvppO"
        role="listbox"
      >
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionOne"
          tabindex="-1"
          value="ValueOne"
        >
          Option One
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionTwo"
          tabindex="-1"
          value="ValueTwo"
        >
          Second Option
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionThree"
          tabindex="-1"
          value="ValueThree"
        >
          Option Three
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`DropDownGroup renders variant 1 1`] = `
<div
  data-testid="test-outSideComponent"
>
  <div>
    something
  </div>
  <div
    aria-haspopup="listbox"
    class="sc-htpNat bEXorH"
    data-testid="test-dropContainer"
    label="Sort By:"
    placeholder="Select An Option"
    tabindex="0"
  >
    <label
      class="dropdown--no-border sc-bdVaJa jOVSaQ"
    >
      <div
        class="dropdown--selected dropdown--no-border sc-ifAKCX dIQjbB"
      >
        Sort By: null
      </div>
      <svg
        class="dropdown--no-border sc-bxivhb bfTkLM"
        fill="currentColor"
        height="12"
        viewBox="0 -1 15 7"
        width="12"
      >
        <path
          d="M13.974.132a.614.614 0 0 0-.762 0L7.17 5.341 1.12.132C.916-.044.563-.037.376.125a.398.398 0 0 0-.167.332c0 .058.012.116.036.17a.385.385 0 0 0 .12.155l6.427 5.54a.568.568 0 0 0 .378.135.568.568 0 0 0 .377-.135l6.427-5.54a.422.422 0 0 0 .156-.325.42.42 0 0 0-.156-.325z"
        />
      </svg>
    </label>
    <div
      class="dropdown__items sc-bwzfXH ewSIUN"
    >
      <div
        class="sc-EHOje hRvppO"
        role="listbox"
      >
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionOne"
          tabindex="-1"
          value="ValueOne"
        >
          Option One
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionTwo"
          tabindex="-1"
          value="ValueTwo"
        >
          Second Option
        </span>
        <span
          class="sc-bZQynM kkZNKr"
          data-testid="test-dropDownOptionThree"
          tabindex="-1"
          value="ValueThree"
        >
          Option Three
        </span>
      </div>
    </div>
  </div>
</div>
`;
